From 900c8325d0c56e79998858f69d3b78994a52607a Mon Sep 17 00:00:00 2001
From: Andrew Zaborowski <andrew.zaborowski@intel.com>
Date: Wed, 9 Apr 2014 11:49:11 -0700
Subject: [PATCH] fs/pstore: Only save one known-to-be-fatal event to

Re-re-apply 98363c5ea0e17cf2e7b776671b99e2dc172a9f8e from kernel-uefi aka.
6bc53fc0a7a63423d41910acb3dde52cef36e741 in kernel-starpeak.

Make sure we don't run out of space in EFI variables too quickly by
making sure that per each boot we only save a single crash dump.  For
now that is the first crash dump we see with a known-to-be-terminal
"reason" value: panic or emergency shutdown.

Change-Id: Ie59b085fd1a6cb766d5436e5dd94d935c7628107
Signed-off-by: Andrew Zaborowski <andrew.zaborowski@intel.com>
---
 fs/pstore/platform.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/fs/pstore/platform.c b/fs/pstore/platform.c
index b4898ea..2269d41 100644
--- a/fs/pstore/platform.c
+++ b/fs/pstore/platform.c
@@ -289,6 +289,17 @@ static void pstore_dump(struct kmsg_dumper *dumper,
 	unsigned long	flags_extra;
 	struct pstore_extra_dumper *extra_dumper;
 
+	/* Don't use pstore space for non-fatal events */
+	if (reason != KMSG_DUMP_PANIC && reason != KMSG_DUMP_EMERG)
+		return;
+
+	/*
+	 * Only dump the first instance of an event, ignore nested
+	 * crashes and similar.
+	 */
+	if (oopscount > 0)
+		return;
+
 	why = get_reason_str(reason);
 
 	if (pstore_cannot_block_path(reason)) {
-- 
1.8.3.2

